sound/soc/sdw_utils/soc_sdw_utils.c

```diff
@@ -1122,6 +1128,47 @@ int asoc_sdw_parse_sdw_endpoints(struct snd_soc_card *card,
 				if (dai_info->quirk && !(dai_info->quirk & ctx->mc_quirk))
 					continue;
 
+				dlc.dai_name = dai_info->dai_name;
+				codec_dai = snd_soc_find_dai_with_mutex(&dlc);
+				if (!codec_dai) {
+					dev_err(dev, "codec dai %s not registered yet\n",
+						dlc.dai_name);
+					return -EAGAIN;
```

Comment:
can this happen?

I am wondering if the deferred-probe mechanism doesn't already guarantee that if this function is reached than all resources were registered.

Probably a question for @charleskeepax and @shumingfan as well.

Reply:
I think these can indeed be missing here. Until snd_soc_register_card (which we are before since we are building the DAI links) is called there is no connection between the drivers, but once that is called ASoC will check if all the components are probed.

Reply:
We don't need the connection. What we need is devm_snd_soc_register_component() being called by codec drivers. The thing is machine driver could probe before codec driver. And machine driver will probe again after codec driver is probed. Not sure what should I do here in this scenario. Maybe just create the dai link as usual and let ASoC do machine probe again?

Reply:
Hmm... yes one needs to be careful here, because the codec probe could be totally asynchronous to us here. So it would be possible that this code here runs, then the codec probes, then we call devm_snd_soc_register_component. So I don't think we could rely on register component to probe defer.

I suspect we need to handle probe defers here, should we not be returning -EPROBE_DEFER rather than -EAGAIN?

Reply:
> Hmm... yes one needs to be careful here, because the codec probe could be totally asynchronous to us here. So it would be possible that this code here runs, then the codec probes, then we call devm_snd_soc_register_component. So I don't think we could rely on register component to probe defer.
> 
> I suspect we need to handle probe defers here, should we not be returning -EPROBE_DEFER rather than -EAGAIN?

Yes, it is for this purpose. Thanks.